linalg/tensor.hpp

```diff
@@ -0,0 +1,2051 @@
+      }
+   }
+   return AB;
+}
+
+/**
+ * @brief this is a shorthand for dot(A, B)
+ */
+template <typename S, typename T, int... m, int... n> MFEM_HOST_DEVICE
+auto operator*(const tensor<S, m...>& A, const tensor<T, n...>& B) ->
+decltype(dot(A, B))
+{
+   return dot(A, B);
+}
+
+/**
+ * @brief Returns the squared Frobenius norm of the tensor
+ * @param[in] A The tensor to obtain the squared norm from
+ */
+template <typename T, int m> MFEM_HOST_DEVICE
+T sqnorm(const tensor<T, m>& A)
+{
+   T total{};
+   for (int i = 0; i < m; i++)
+   {
+      total += A[i] * A[i];
+   }
+   return total;
+}
+
+/// @overload
+template <typename T, int m, int n> MFEM_HOST_DEVICE
+T sqnorm(const tensor<T, m, n>& A)
+{
+   T total{};
+   for (int i = 0; i < m; i++)
+   {
+      for (int j = 0; j < n; j++)
+      {
+         total += A[i][j] * A[i][j];
+      }
+   }
+   return total;
+}
+
+// /// @overload
+/**
+ * @brief Returns the Frobenius norm of the tensor
+ * @param[in] A The tensor to obtain the norm from
+ */
+template <typename T, int... n> MFEM_HOST_DEVICE
+T norm(const tensor<T, n...>& A)
+{
+   return std::sqrt(sqnorm(A));
+}
+
+/**
+ * @brief Normalizes the tensor
+ * Each element is divided by the Frobenius norm of the tensor, @see norm
+ * @param[in] A The tensor to normalize
+ */
+template <typename T, int... n> MFEM_HOST_DEVICE
+auto normalize(const tensor<T, n...>& A) ->
+decltype(A / norm(A))
+{
+   return A / norm(A);
+}
+
+/**
+ * @brief Returns the trace of a square matrix
+ * @param[in] A The matrix to compute the trace of
+ * @return The sum of the elements on the main diagonal
+ */
+template <typename T, int n> MFEM_HOST_DEVICE
+T tr(const tensor<T, n, n>& A)
+{
+   T trA{};
+   for (int i = 0; i < n; i++)
+   {
+      trA = trA + A[i][i];
+   }
+   return trA;
+}
+
+/**
+ * @brief Returns the symmetric part of a square matrix
+ * @param[in] A The matrix to obtain the symmetric part of
+ * @return (1/2) * (A + A^T)
+ */
+template <typename T, int n> MFEM_HOST_DEVICE
+tensor<T, n, n> sym(const tensor<T, n, n>& A)
+{
+   tensor<T, n, n> symA{};
+   for (int i = 0; i < n; i++)
+   {
+      for (int j = 0; j < n; j++)
+      {
+         symA[i][j] = 0.5 * (A[i][j] + A[j][i]);
+      }
+   }
+   return symA;
+}
+
+/**
+ * @brief Calculates the deviator of a matrix (rank-2 tensor)
+ * @param[in] A The matrix to calculate the deviator of
+ * In the context of stress tensors, the deviator is obtained by
+ * subtracting the mean stress (average of main diagonal elements)
+ * from each element on the main diagonal
+ */
+template <typename T, int n> MFEM_HOST_DEVICE
+tensor<T, n, n> dev(const tensor<T, n, n>& A)
+{
+   auto devA = A;
+   auto trA  = tr(A);
+   for (int i = 0; i < n; i++)
+   {
+      devA[i][i] -= trA / n;
+   }
+   return devA;
+}
+
+/**
+ * @brief Obtains the identity matrix of the specified dimension
+ * @return I_dim
+ */
+template <int dim>
+MFEM_HOST_DEVICE tensor<double, dim, dim> Identity()
```

Comment:
Why is this one capitalized, and all the rest are lowercase?

What is the convention here for capitalizing function names? In general in MFEM we capitalize function names, but I understand that for generic programming and compatibility with the standard library, some functions must not be captialized (e.g. `get`).

Reply:
It makes sense to call this `identity()` since all other functions and classes here use lowercase.